### ISSUE TITLE
Support GPM

### ISSUE THREAD

**Issue by secrettriangle** [chjj/blessed#94]
http://en.wikipedia.org/wiki/GPM_%28software%29

**Comment by mgliewe** [comment]
I was thinking the same thing. could be usefull.

could use ttyname() (unsupported in nodes lib) or ```sytem("tty")``` to find out wheter on console, a check on $TERM='linux' could suffice as well.

would have to open extra socket to gpm and either inject into regular input stream or just emit events to program.

**Comment by chjj** [comment]
I tried to implement as many mouse protocols as possible. I stopped at GPM because I couldn't find a specification for the protocol at the time. 

Is there documentation anywhere, or is it just in code? Documentation would be very helpful. Otherwise we'll have to dig into ncurses again, and gpm itself.

**Comment by mgliewe** [comment]
afaik you'll have to connect to an extra /dev/xx socket and listen for events; i have some links about that somewhere, will report back soon when i got time to look it up.

**Comment by mgliewe** [comment]
I promised to come back to you on this issue; alas i couldn't find what i was looking for. must have been gone over time, havent used this kinda stuff for a while.

what i can remember from old times: 
- one has to connect to /dev/gpmdata; unsure if via open or as a socket, or if it even matters?
- protocol is basically PS2
- copy of gpm GIT is @ https://github.com/telmich/gpm (found this one on google)
- i guess of special interrest ist https://github.com/telmich/gpm/blob/master/src/headers/gpm-proto.h

sorry for not beeing so usefull this time, still somewhat busy, but awaiting some free time and i'm eager to play around somewhat with blessed as soon as i get some time spare.

**Comment by piranna** [comment]
I find the difficult part to inject the ANSI sequences of the mouse events at stdio stream, once that's possible, it would not be needed support for GPM at all since Linux terminals and PTYs would work the same as X terminals (xterm & friends). I think this is the most flexible option.

**Comment by mgliewe** [comment]
did some hacking on gpm, this code reports gpm events on my linux console.

@chjj: isn't injecting into  blessed event stream  as simple as ```program.emit('data', 'some-ansi-escapes')```?

```
var net=require('net');
var fs=require('fs');
var EventEmitter=require('events').EventEmitter;
var util = require('util')

var GPM_USE_MAGIC=false;
var GPM_MOVE=1, GPM_DRAG=2, GPM_DOWN=4, GPM_UP=8, GPM_DOUBLE=32,GPM_MFLAG=128;

function GpmClient(options) {
	if (!this instanceof GpmClient) {
		return new GpmClient(options);
	}
	
	EventEmitter.call(this);
	//#define GPM_MAGIC 0x47706D4C /* "GpmL" */
	var GPM_MAGIC=0x47706D4C;
	var GPM_SOCKET="/dev/gpmctl";
	
	var pid=process.pid;
	var tty=/tty[0-9]+$/.exec(fs.readlinkSync('/proc/'+pid+'/fd/0'));
	var vc;
	if (tty) {
		var tty=tty[0];
		vc=/[0-9]+$/.exec(tty)[0];
		console.log(tty, vc);
	}
	
	/*
		typedef struct Gpm_Connect {
	  		unsigned short eventMask, defaultMask;
	  		unsigned short minMod, maxMod;
	  		int pid;
	  		int vc;
		} Gpm_Connect;
	*/	
	function send_config(sock, Gpm_Connect) {
		if (GPM_USE_MAGIC) {
			var buffer=new Buffer(20);
			buffer.writeUInt32LE(GPM_MAGIC, 0);
			buffer.writeUInt16LE(Gpm_Connect.eventMask, 4);
			buffer.writeUInt16LE(Gpm_Connect.defaultMask, 6);
			buffer.writeUInt16LE(Gpm_Connect.minMod, 8);
			buffer.writeUInt16LE(Gpm_Connect.maxMod, 10);
			buffer.writeInt16LE(process.pid, 12);
			buffer.writeInt16LE(Gpm_Connect.vc, 16);
		} else {
			var buffer=new Buffer(16);
			buffer.writeUInt16LE(Gpm_Connect.eventMask, 0);
			buffer.writeUInt16LE(Gpm_Connect.defaultMask, 2);
			buffer.writeUInt16LE(Gpm_Connect.minMod, 4);
			buffer.writeUInt16LE(Gpm_Connect.maxMod, 6);
			buffer.writeInt16LE(pid, 8);
			buffer.writeInt16LE(vc, 12);
		}
		sock.write(buffer, function () { console.log('sent config');});
	}   
	
	/* typedef struct Gpm_Event {
		  unsigned char buttons, modifiers;  // try to be a multiple of 4 
		  unsigned short vc;
		  short dx, dy, x, y; // displacement x,y for this event, and absolute x,y 
		  enum Gpm_Etype type;
		  // clicks e.g. double click are determined by time-based processing 
		  int clicks;
		  enum Gpm_Margin margin;
		  // wdx/y: displacement of wheels in this event. Absolute values are not 
		  // required, because wheel movement is typically used for scrolling
		  // or selecting fields, not for cursor positioning. The application
		  // can determine when the end of file or form is reached, and not
		  // go any further. 
		  // A single mouse will use wdy, "vertical scroll" wheel. 
		  short wdx, wdy;
		} Gpm_Event;
		*/
	function parseEvent(raw) {
		var evnt={};
		evnt.buttons=raw[0];
		evnt.modifiers=raw[1];
		evnt.vc=raw.readUInt16LE(2);
		evnt.dx=raw.readInt16LE(4);
		evnt.dy=raw.readInt16LE(6);
		evnt.x=raw.readInt16LE(8);
		evnt.y=raw.readInt16LE(10);
		evnt.type=raw.readInt16LE(12);
		evnt.clicks=raw.readInt32LE(16);
		evnt.margin=raw.readInt32LE(20);
		evnt.wdx=raw.readInt16LE(24);
		evnt.wdy=raw.readInt16LE(26);
		return evnt;
	}
	
	
	var conf =  {
		eventMask: 0xffff, 
		defaultMask: 0xffff,
		minMod: 0,  
		maxMod: 0xffff 
	};

	if (tty) {
		var gpm=net.createConnection(GPM_SOCKET);
		
		gpm.on('connect', function() {
			console.log('connected to gpm');
			send_config(gpm, conf);
			
		})
		var self=this;
		gpm.on('data', function(packet) {
			var evnt=parseEvent(packet);
			switch(evnt.type & 15) {
			case GPM_MOVE:	
				if (evnt.dx || evnt.dy) self.emit('move', evnt.buttons, evnt.modifiers, evnt.x, evnt.y)
				if (evnt.wdx || evnt.wdy) self.emit('mousewheel', evnt.buttons, evnt.modifiers, evnt.wdx, evnt.wdy)
				break;
			case GPM_DRAG:	
				if (evnt.dx || evnt.dy) self.emit('drag', evnt.buttons, evnt.modifiers, evnt.x, evnt.y)
				if (evnt.wdx || evnt.wdy) self.emit('mousewheel', evnt.buttons, evnt.modifiers, evnt.wdx, evnt.wdy)
				break;
			case GPM_DOWN:	
				self.emit('btndown', evnt.buttons, evnt.modifiers, evnt.x, evnt.y)
				if (evnt.type & GPM_DOUBLE) {
					self.emit('dblclick', evnt.buttons, evnt.modifiers, evnt.x, evnt.y)
				}
			break;
			case GPM_UP:	
				self.emit('btnup', evnt.buttons, evnt.modifiers, evnt.x, evnt.y)
				if (!(evnt.type & GPM_MFLAG)) {
					self.emit('click', evnt.buttons, evnt.modifiers, evnt.x, evnt.y)
				}
			break;				
			}
		})
		gpm.on('error', function(err) {
			console.log('GPM ERROR', err);
		})
	}
}

util.inherits(GpmClient, EventEmitter);



c=new GpmClient();
c.on('click', function(btn, mask, x, y) { console.log('click '+btn+'@('+x+','+y+')'); });
c.on('btnup', function(btn, mask, x, y) { console.log('btnup '+btn+'@('+x+','+y+')'); });
c.on('btndown', function(btn, mask, x, y) { console.log('btndown '+btn+'@('+x+','+y+')'); });
c.on('dblclick', function(btn, mask, x, y) { console.log('dblclick '+btn+'@('+x+','+y+')'); });
c.on('move', function(btn, mask, x, y) { console.log('move '+btn+'@('+x+','+y+')'); });
c.on('drag', function(btn, mask, x, y) { console.log('drag '+btn+'@('+x+','+y+')'); });
c.on('mousewheel', function(btn, mask, wx, wy) { console.log('scroll '+btn+'@('+wx+','+wy+')'); });

```

**Comment by mgliewe** [comment]
i fiddled a gpm client into blessed  program.js. will work on linux/little endian only for now.

Would  anybody  care  to give it a try and do some testing on his local clone, before i issue a pull request?

https://github.com/mgliewe/blessed.git

**Comment by piranna** [comment]
I still think is a better aproach to write an independent client program
that write the ANSI codes on the stdin stream, but this is not a bad job,
though :-) The GPM socket is hardcoded, isn't it? How could be done to set
an alternative one? How could be changed to also allow to listen directly
to the mouse device (I'm thinking about having several mouses attached to a
computer and being able to control them separately, that I think GPM is not
capable of...)?

2015-01-28 3:47 GMT+01:00 mgliewe <notifications@github.com>:

> i fiddled a gpm client into blessed program.js. will work on linux/little
> endian only for now.
>
> Would anybody care to give it a try and do some testing on his local
> clone, before i issue a pull request?
>
> https://github.com/mgliewe/blessed.git
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/chjj/blessed/issues/94#issuecomment-71772164>.
>



-- 
"Si quieres viajar alrededor del mundo y ser invitado a hablar en un monton
de sitios diferentes, simplemente escribe un sistema operativo Unix."
– Linus Tordvals, creador del sistema operativo Linux

**Comment by mgliewe** [comment]
@piranna: gpm socket path is hardcoded, same on all systems. 
Direkt connection to mice might proof difficult, gpm does all the hard work already. And it provides the hardware cursor, witch might be difficult to archive using ansi emulation.
One can connect gpm to /dev/input/mice, this will allow multiple mice to use.

I'm using the control interface to gpm, which ist different to the ps/gpmdata protocol.

**Comment by chjj** [comment]
@piranna, I have to agree with @mgliewe. What would be the benefit of doing that?

Good job, @mgliewe. I'm awaiting your PR.

**Comment by piranna** [comment]
> @piranna, I have to agree with @mgliewe. What would be the benefit of doing that?

As I said, being able to plug several mouses to the computer and being
able to control them separately, by diferent apps on several consoles,
or several users on the same app (like games)... I'm not sure if GPM
is capable to do this.

On a more personal note, being able users on NodeOS to access directly
to the mouse devices and not need an external, root priviledges
process to access to them, specially if they will be shared by
everybody. I'm looking for other more generic alternatives (like
injecting the ANSI codes on stdin, as I said), though...



-- 
"Si quieres viajar alrededor del mundo y ser invitado a hablar en un
monton de sitios diferentes, simplemente escribe un sistema operativo
Unix."
– Linus Tordvals, creador del sistema operativo Linux

**Comment by chjj** [comment]
@mgliewe, I'm pretty eager to get this PR in so we can finally be done with mouse protocols. :)

**Comment by chjj** [comment]
Implemented with #101.
